File: example-crs-webservice/crs-userspace/libs/libMSA/cpp/src/libmsa/cpu/affinity.cpp
#include "libmsa/cpu/affinity.h"
#include <pthread.h>
#include <unistd.h>
#include <sched.h>
#include <sys/syscall.h>

void CpuAffinity::setCpuAffinity(const std::vector<int>& cpus) {
    pid_t pid = getpid();

    cpu_set_t cpuset;
    CPU_ZERO(&cpuset);
    for (int cpu : cpus) {
        CPU_SET(cpu, &cpuset);
    }

    sched_setaffinity(pid, sizeof(cpu_set_t), &cpuset);
}

File: example-crs-webservice/crs-userspace/libs/libMSA/go/cpu/affinity.go
package cpu

import (
	"fmt"
	"os"
	"golang.org/x/sys/unix"
)

// setCPUAffinity sets the CPU affinity for the current process
func SetCPUAffinity(cpus []int) error {
	// Get the current process ID
	pid := os.Getpid()

	// Create a CPU mask
	var mask unix.CPUSet
	for _, cpu := range cpus {
		mask.Set(cpu)
	}

	// Set CPU affinity for the process
	if err := unix.SchedSetaffinity(pid, &mask); err != nil {
		return fmt.Errorf("failed to set CPU affinity for process %d: %v", pid, err)
	}

	return nil
}